File: tests/integration/test_constrained_types.py
"""Integration tests for constrained ASN.1 types."""
import pytest

def compile_and_import(tmp_path):
    import subprocess, shutil, sys
    asn_file = tmp_path / "2000_constrained_types.asn1"
    shutil.copy("tests/vectors/asn1/2000_constrained_types.asn1", asn_file)
    out_dir = tmp_path / "out"
    out_dir.mkdir()
    runtime_dir = tmp_path / "asnvil_runtime"
    shutil.copytree("asnvil-runtime-python", runtime_dir)
    result = subprocess.run(
        [sys.executable, "-m", "cargo", "run", "--", "-o", str(out_dir), str(asn_file)],
        cwd=".", capture_output=True, text=True
    )
    if result.returncode != 0:
        # Try direct cargo run
        result = subprocess.run(
            ["cargo", "run", "--", "-o", str(out_dir), str(asn_file)],
            cwd=".", capture_output=True, text=True
        )
        assert result.returncode == 0, f"Compilation failed: {result.stderr}"
    import importlib.util
    spec = importlib.util.spec_from_file_location("ConstrainedTypes", out_dir / "ConstrainedTypes.py")
    mod = importlib.util.module_from_spec(spec)
    sys.modules["ConstrainedTypes"] = mod
    spec.loader.exec_module(mod)
    return mod


def test_valid_user_record(tmp_path):
    mod = compile_and_import(tmp_path)
    user = mod.UserRecord(id=42, name="Alice", age=30)
    data = user.encode_der()
    decoded = mod.UserRecord.decode_der(data)
    assert decoded.id == 42
    assert decoded.name == "Alice"
    assert decoded.age == 30


def test_valid_user_with_optional(tmp_path):
    mod = compile_and_import(tmp_path)
    user = mod.UserRecord(id=100, name="Bob", age=25, status=3, notes="Test notes")
    data = user.encode_der()
    decoded = mod.UserRecord.decode_der(data)
    assert decoded.id == 100
    assert decoded.name == "Bob"
    assert decoded.status == 3
    assert decoded.notes == "Test notes"


def test_valid_boundary_id_zero(tmp_path):
    mod = compile_and_import(tmp_path)
    user = mod.UserRecord(id=0, name="X", age=0)
    data = user.encode_der()
    decoded = mod.UserRecord.decode_der(data)
    assert decoded.id == 0


def test_valid_boundary_id_max(tmp_path):
    mod = compile_and_import(tmp_path)
    user = mod.UserRecord(id=1000, name="Y", age=150)
    data = user.encode_der()
    decoded = mod.UserRecord.decode_der(data)
    assert decoded.id == 1000
    assert decoded.age == 150


def test_valid_name_min_length(tmp_path):
    mod = compile_and_import(tmp_path)
    user = mod.UserRecord(id=1, name="A", age=1)
    data = user.encode_der()
    decoded = mod.UserRecord.decode_der(data)
    assert decoded.name == "A"


def test_valid_name_max_length(tmp_path):
    mod = compile_and_import(tmp_path)
    user = mod.UserRecord(id=1, name="A" * 50, age=1)
    data = user.encode_der()
    decoded = mod.UserRecord.decode_der(data)
    assert len(decoded.name) == 50


def test_invalid_id_exceeds_max(tmp_path):
    mod = compile_and_import(tmp_path)
    from asnvil_runtime import ConstraintViolationError
    user = mod.UserRecord(id=1001, name="Test", age=25)
    with pytest.raises(ConstraintViolationError):
        user.encode_der()


def test_invalid_age_negative(tmp_path):
    mod = compile_and_import(tmp_path)
    from asnvil_runtime import ConstraintViolationError
    user = mod.UserRecord(id=1, name="Test", age=-1)
    with pytest.raises(ConstraintViolationError):
        user.encode_der()


def test_invalid_name_empty(tmp_path):
    mod = compile_and_import(tmp_path)
    from asnvil_runtime import ConstraintViolationError
    user = mod.UserRecord(id=1, name="", age=25)
    with pytest.raises(ConstraintViolationError):
        user.encode_der()

File: tests/vectors/constrained_types/test_constrained_types.rs
//! ```cargo
//! [dependencies]
//! asnvil-runtime-rust = { path = "__REPO_ROOT__/asnvil-runtime-rust" }
//! num-bigint = "0.4"
//! ```

#[path = "./ConstrainedTypes.rs"]
mod generated;

use generated::UserRecord;
use num_bigint::BigInt;
use asnvil_runtime_rust::ConstraintViolationError;

#[test]
fn test_valid_user_record() {
    let user = UserRecord {
        id: BigInt::from(42),
        name: "Alice".to_string(),
        age: BigInt::from(30),
        status: None,
        notes: None,
    };
    let data = user.encode_der().expect("Failed to encode");
    let decoded = UserRecord::decode_der(&data).expect("Failed to decode");
    assert_eq!(decoded.id, BigInt::from(42));
    assert_eq!(decoded.name, "Alice");
    assert_eq!(decoded.age, BigInt::from(30));
}

#[test]
fn test_valid_user_with_optional() {
    let user = UserRecord {
        id: BigInt::from(100),
        name: "Bob".to_string(),
        age: BigInt::from(25),
        status: Some(BigInt::from(3)),
        notes: Some("Test notes".to_string()),
    };
    let data = user.encode_der().expect("Failed to encode");
    let decoded = UserRecord::decode_der(&data).expect("Failed to decode");
    assert_eq!(decoded.id, BigInt::from(100));
    assert_eq!(decoded.name, "Bob");
    assert_eq!(decoded.status, Some(BigInt::from(3)));
    assert_eq!(decoded.notes, Some("Test notes".to_string()));
}

#[test]
fn test_valid_boundary_id_zero() {
    let user = UserRecord {
        id: BigInt::from(0),
        name: "X".to_string(),
        age: BigInt::from(0),
        status: None,
        notes: None,
    };
    let data = user.encode_der().expect("Failed to encode");
    let decoded = UserRecord::decode_der(&data).expect("Failed to decode");
    assert_eq!(decoded.id, BigInt::from(0));
}

#[test]
fn test_valid_boundary_id_max() {
    let user = UserRecord {
        id: BigInt::from(1000),
        name: "Y".to_string(),
        age: BigInt::from(150),
        status: None,
        notes: None,
    };
    let data = user.encode_der().expect("Failed to encode");
    let decoded = UserRecord::decode_der(&data).expect("Failed to decode");
    assert_eq!(decoded.id, BigInt::from(1000));
    assert_eq!(decoded.age, BigInt::from(150));
}

#[test]
fn test_valid_name_min_length() {
    let user = UserRecord {
        id: BigInt::from(1),
        name: "A".to_string(),
        age: BigInt::from(1),
        status: None,
        notes: None,
    };
    let data = user.encode_der().expect("Failed to encode");
    let decoded = UserRecord::decode_der(&data).expect("Failed to decode");
    assert_eq!(decoded.name, "A");
}

#[test]
fn test_valid_name_max_length() {
    let user = UserRecord {
        id: BigInt::from(1),
        name: "A".repeat(50),
        age: BigInt::from(1),
        status: None,
        notes: None,
    };
    let data = user.encode_der().expect("Failed to encode");
    let decoded = UserRecord::decode_der(&data).expect("Failed to decode");
    assert_eq!(decoded.name.len(), 50);
}

#[test]
fn test_invalid_id_exceeds_max() {
    let user = UserRecord {
        id: BigInt::from(1001),
        name: "Test".to_string(),
        age: BigInt::from(25),
        status: None,
        notes: None,
    };
    assert!(user.encode_der().is_err(), "Expected ConstraintViolationError for id > 1000");
}

#[test]
fn test_invalid_age_negative() {
    let user = UserRecord {
        id: BigInt::from(1),
        name: "Test".to_string(),
        age: BigInt::from(-1),
        status: None,
        notes: None,
    };
    assert!(user.encode_der().is_err(), "Expected ConstraintViolationError for age < 0");
}

#[test]
fn test_invalid_name_empty() {
    let user = UserRecord {
        id: BigInt::from(1),
        name: "".to_string(),
        age: BigInt::from(25),
        status: None,
        notes: None,
    };
    assert!(user.encode_der().is_err(), "Expected ConstraintViolationError for empty name");
}